os/kernel/semaphore/sem_destroy.c

```diff
@@ -85,6 +88,26 @@
 /****************************************************************************
  * Private Functions
  ****************************************************************************/
+#ifdef CONFIG_APP_BINARY_SEPARATION
+extern sq_queue_t g_sem_list;
+
+static void sem_unregister(FAR sem_t *sem)
+{
+	sem_t *sem_ptr;
+	irqstate_t flags;
+
+	sem_ptr = (sem_t *)sq_peek(&g_sem_list);
+	while (sem_ptr) {
+		if (sem_ptr == sem) {
+			flags = irqsave();
+			sq_rem((FAR sq_entry_t *)sem, &g_sem_list);
```

Comment:
I think no need to traverse through entire list. That logic is taken care in sq_rem.